src/librustc_mir/const_eval/eval_queries.rs

```diff
@@ -128,10 +131,16 @@ pub(super) fn op_to_const<'tcx>(
         op.try_as_mplace(ecx)
     };
 
+    let alloc_map = ecx.tcx.alloc_map.lock();
+    let try_unwrap_memory = |alloc_id: AllocId| match alloc_map.get(alloc_id) {
+        Some(GlobalAlloc::Memory(mem)) => Ok(mem),
+        Some(GlobalAlloc::Static(_)) => Err(RefersToStatic),
```

Comment:
This breaks our assumption, that any constant of scalar type or `&str`/`&[u8]` type is `ConstValue::Scalar` or `ConstValue::Slice` respectively. If such a constant is used in pattern matching, things will break in bad ways (e.g. two identical patterns not comparing equal). I guess such constants must be forbidden from occurring in patterns and const generics.

Reply:
Oh... pattern matching assumes that references in consts can only point to immutable memory? That's... a rather big assumption! It was true so far but only due to extreme conservatism on the side of our static const checks.

However, I feel like the fact that `&str` is an immutable shared reference should count for something. Maybe interning should ensure that shared references only point to read-only allocations, or so? (recursively)

Reply:
If the `&str` came from a `&mut str`, we can't turn the `&mut str` allocation into an immutable one. And if we allow constants to refer to statics, that can happen, and we must keep the pointer to the static, because otherwise we violate the fact that a pointer to the same static must have the same address.

Reply:
How'd that look like though? When a const takes an `&mut` to something, that reference has lifetime `'static`, so nothing else ever in the entire program is allowed to access the memory behind that unique reference.

Though when we have `&i32`, that could indeed be from an `&mut i32` that only covers part of an allocation and the rest of it could still permit mutation.

I think allowing references in consts in patterns was a big mistake... do we permit this for all types or just `&str`/`&[u8]`?

Reply:
> I think allowing references in consts in patterns was a big mistake... do we permit this for all types or just &str/&[u8]?

Hold up :D we need special rules for pattern constants and const generics anyway. @eddyb talked to me about this already. E.g. we can't non-normalized  padding or non-interned references in constants used in const generics. We do need a separate interning & normalization for such constants anyway. Not sure how we can detect references into statics though, maybe we need a new flag in `Alloation` that marks the allocation as part of a static, and thus having a unique address.

Reply:
I opened https://github.com/rust-lang/const-eval/issues/42 for the soundness concerns.

We probably have to get that answered before allowing consts pointing to mutable statics in our dynamic checks. Sorry @rpjohnst, I was entirely unaware of this particular minefield.

For now, could you remove those changes from this PR? I.e., dynamically permit consts pointing to *immutable* statics, but make no effort to permit consts pointing to *mutable* statics?

Reply:
As long as we do it behind a feature gate, that's perfectly fine in my book.

Reply:
@rpjohnst's use-case doesn't need consts pointing to mutable statics though. I think this change is part of this PR because I recommended them to do that, because I thought we could weaken our dynamic checks all the way to "just prevent reads from mutable statics, everything else is fine".

I was wrong as I didn't know about pattern-matching-soundness. In light of this new information, I'd prefer it we didn't weaken our dynamic checks quite so far.

Reply:
Sure, will do. To clarify- do we want these dynamic checks behind a feature gate as well? Or can we still rely on the static check and just feature gate any changes there?

Reply:
If you avoid touching validity (beyond what is necessary to enable consts to point to immutable statics), I don't think any feature gates are needed for weakening the dynamic checks.